tutorials/Tutorial7_generator.py

```diff
@@ -0,0 +1,70 @@
+    ),
+    Document(
+        text="""Schopenhauer, describing him as an ultimately shallow thinker: ""Schopenhauer has quite a crude mind ... where real depth starts, his comes to an end."" His friend Bertrand Russell had a low opinion on the philosopher, and attacked him in his famous ""History of Western Philosophy"" for hypocritically praising asceticism yet not acting upon it. On the opposite isle of Russell on the foundations of mathematics, the Dutch mathematician L. E. J. Brouwer incorporated the ideas of Kant and Schopenhauer in intuitionism, where mathematics is considered a purely mental activity, instead of an analytic activity wherein objective properties of reality are""",
+        meta={"name": "1"}
+    ),
+    Document(
+        text="""The Dothraki vocabulary was created by David J. Peterson well in advance of the adaptation. HBO hired the Language Creatio""",
+        meta={"name": "2"}
+    ),
+    Document(
+        text="""The title of the episode refers to the Great Sept of Baelor, the main religious building in King's Landing, where the episode's pivotal scene takes place. In the world created by George R. R. Martin""",
+        meta={}
+    )
+]
+
+document_store = FAISSDocumentStore(faiss_index_factory_str="HNSW")
+document_store.delete_all_documents()
+document_store.write_documents(documents)
+
+retriever = DensePassageRetriever(document_store=document_store,
+                                  query_embedding_model="facebook/dpr-question_encoder-single-nq-base",
+                                  passage_embedding_model="facebook/dpr-ctx_encoder-single-nq-base",
+                                  use_gpu=False, embed_title=True,
+                                  remove_sep_tok_from_untitled_passages=True)
+
+document_store.update_embeddings(retriever=retriever)
+
+docs_with_emb = document_store.get_all_documents(index="test_dpr")
+
+res = retriever.retrieve(query="Which philosopher attacked Schopenhauer?", top_k=1)
+
+# Generator part
+
+tokenizer = RagTokenizer.from_pretrained("facebook/rag-sequence-base")
+model = RagSequenceForGeneration.from_pretrained("facebook/rag-sequence-base")
+
+input_dict = tokenizer.prepare_seq2seq_batch("Which philosopher attacked Schopenhauer?", return_tensors="pt")
+print("input_dict: ", input_dict)
+
+doc_dict = tokenizer.prepare_seq2seq_batch(res[0].text, return_tensors="pt")
+
+question_hidden_states = model.question_encoder(input_dict["input_ids"])[0]
+
+#doc_scores = torch.bmm(question_hidden_states.unsqueeze(1),
+#                       res[0].embedding).squeeze(1)
+
```

Comment:
@tholor can you please help here. I am trying to apply following example from transformers but using base model https://huggingface.co/facebook/rag-sequence-base -
```
>>> from transformers import RagTokenizer, RagRetriever, RagSequenceForGeneration
>>> import torch

>>> tokenizer = RagTokenizer.from_pretrained("facebook/rag-sequence-nq")
>>> retriever = RagRetriever.from_pretrained("facebook/rag-sequence-nq", index_name="exact", use_dummy_dataset=True)
>>> # initialize with RagRetriever to do everything in one forward call
>>> model = RagSequenceForGeneration.from_pretrained("facebook/rag-token-nq", retriever=retriever)

>>> input_dict = tokenizer.prepare_seq2seq_batch("How many people live in Paris?", "In Paris, there are 10 million people.", return_tensors="pt")
>>> input_ids = input_dict["input_ids"]
>>> outputs = model(input_ids=input_ids, labels=input_dict["labels"])

>>> # or use retriever seperately
>>> model = RagSequenceForGeneration.from_pretrained("facebook/rag-sequence-nq", use_dummy_dataset=True)
>>> # 1. Encode
>>> question_hidden_states = model.question_encoder(input_ids)[0]
>>> # 2. Retrieve
>>> docs_dict = retriever(input_ids.numpy(), question_hidden_states.detach().numpy(), return_tensors="pt")
>>> doc_scores = torch.bmm(question_hidden_states.unsqueeze(1), docs_dict["retrieved_doc_embeds"].float().transpose(1, 2)).squeeze(1)
>>> # 3. Forward to generator
>>> outputs = model(context_input_ids=docs_dict["context_input_ids"], context_attention_mask=docs_dict["context_attention_mask"], doc_scores=doc_scores, decoder_input_ids=input_dict["labels"])

>>> # or directly generate
>>> generated = model.generate(input_ids=input_dict["input_ids"])
>>> generated_string = tokenizer.batch_decode(generated, skip_special_tokens=True)
```